File: KimJiYu/week_7/boj_1010.java
package org.example.nestnet_algorithm_2024_2_beginner.KimJiYu.week_7;

import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt(); // 테스트 케이스 수

        StringBuilder sb = new StringBuilder();
        while (T-- > 0) {
            int N = sc.nextInt(); // 다리를 놓을 사이트 개수
            int M = sc.nextInt(); // 강 서쪽의 사이트 개수

            sb.append(combination(M, N)).append("\n");
        }

        System.out.print(sb);
    }

    // 이항 계수 계산 함수
    public static long combination(int n, int r) {
        if (r > n) return 0; // 불가능한 경우
        if (r == 0 || r == n) return 1; // 조합의 경계 조건

        r = Math.min(r, n - r); // nCr = nC(n-r) 대칭성 활용
        long result = 1;

        for (int i = 0; i < r; i++) {
            result *= (n - i);
            result /= (i + 1);
        }

        return result;
    }
}

File: ChaeJangWoo/week_7/boj_1010.cpp
#include<iostream>
using namespace std;

// 조합 계산 함수 (팩토리얼 직접 계산 대신 최적화)
long long combination(int n, int r) {
    if (r > n / 2) r = n - r; // C(n, r) == C(n, n-r)
    long long result = 1;
    for (int i = 1; i <= r; i++) {
        result *= (n - i + 1);
        result /= i;
    }
    return result;
}

int main() {
    int t;
    cin >> t;
    int n[t], m[t];
    long long rest[t]; // 결과를 저장할 배열
    for (int i = 0; i < t; i++) {
        cin >> n[i] >> m[i];
        rest[i] = combination(m[i], n[i]); // 최적화된 조합 계산
    }
    for (int i = 0; i < t; i++) {
        cout << rest[i] << endl; // 결과 출력
    }
    return 0;
}
